### PR TITLE
feat: add settings tab with language selection

### LINKED ISSUE #21: Create Settings tab
- add language

### PR DESCRIPTION
**Pull request by Painguin**
close #21.

Creates a Settings tab with language selection.

### CODE REVIEW

File: src/components/Settings.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+const ConnectedComponent = connect(
+  mapStateToProps,
+  mapDispatchToProps,
+)(Settings);
+
+const StyledComponent = withStyles(styles)(ConnectedComponent);
+
+const TranslatedComponent = withTranslation()(StyledComponent);
+
+export default withRouter(TranslatedComponent);
```
Comment (pyphilia) [inline src/components/Settings.js:100]: Do you need `withRouter` in this component?

File: src/config/messages.js [hunk 2]
```diff
@@ -14,6 +14,8 @@ const ERROR_SETTING_DATABASE = 'There was an error setting the database';
 const ERROR_GETTING_DATABASE = 'There was an error retrieving the database';
 const ERROR_DELETING_DATASET_MESSAGE =
   'There was an error deleting the dataset';
+const ERROR_GETTING_LANGUAGE = 'There was an error getting the language.';
```
Comment (pyphilia) [inline src/config/messages.js:17]: I think we are going with `retrieving` here.

File: src/langs/fr.json [hunk 3]
```diff
@@ -7,6 +7,22 @@
...
+    "Developer": "Développeur",
+    "Load a File": "Importer un ficher",
+    "Browse": "Parcourir",
+    "Submit": "Envoyer",
+    "Danger Zone! Proceed with caution as changes to this section might lead to data loss.": "Attention ! Les modifications de cette section pourraient entraîner une perte de données.",
+    "The database is empty.": "La base de données est vide.",
+    "Manually Edit the Database": "Modifier Manuellement la Base de Données",
+    "Use Sample Database": "Utiliser une Base de Données Exemple",
+    "Language": "Langue",
+    "KB": "KB",
```
Comment (pyphilia) [inline src/langs/fr.json:24]: Change to `Ko`

**Review by pyphilia — CHANGES_REQUESTED** [review]
Really nice! Here some minor comments.

**Review by pyphilia — APPROVED** [review]

File: src/langs/en.json [hunk 4]
```diff
@@ -7,6 +7,22 @@
...
+    "Unknown": "Unknown",
+    "Load": "Load",
+    "Datasets": "Datasets",
+    "Algorithms": "Algorithms",
+    "Settings": "Settings",
+    "Developer": "Developer",
+    "Load a File": "Load a File",
+    "Browse": "Browse",
+    "Submit": "Submit",
+    "Danger Zone! Proceed with caution as changes to this section might lead to data loss.": "Attention ! Les modifications de cette section pourraient entraîner une perte de données.",
```
Comment (hagoptaminian) [inline src/langs/en.json:19]: This is a French translation in `en.json`. Also, no space between 'Attention' and '!'

**Review by hagoptaminian — CHANGES_REQUESTED** [review]

**Comment by hagoptaminian** [comment]
@Painguin this looks good to me - left a small comment on a mistake in `en.json`. Otherwise, I don't think it's great the way the dropdown menu in the new component stretches across the entire page:

Screen Shot 2020-10-07 at 3 28 13 PM

Maybe giving it width of 25% is better. Also, instead of a Menu item, I think it's nicer to have the settings accessible via a settings icon in the header, which when clicked on pops up a modal through which the user can adjust language (and in the future other) settings (that way they don't have to navigate away from the page they're on). This is similar to how Postman and other apps do it:

Screen Shot 2020-10-07 at 2 02 08 PM

Screen Shot 2020-10-07 at 2 02 43 PM

**Comment by Painguin** [comment]
Thank you for comment @hagoptaminian. I have modified the settings tab into a modal as you suggested.

**Review by hagoptaminian — APPROVED** [review]

File: src/actions/settings.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+          type: GET_LANGUAGE_SUCCESS,
+          payload: lang,
+        });
+      }
+      dispatch(flagGettingLanguage(false));
+    });
+  } catch (e) {
+    // eslint-disable-next-line no-console
+    console.error(e);
+    toastr.error(ERROR_MESSAGE_HEADER, ERROR_GETTING_LANGUAGE);
```
Comment (pyphilia) [inline src/actions/settings.js:38]: You'll probably need `dispatch(flagGettingLanguage(false));` in the `catch` block as well. (same below)

File: src/components/common/Header.js [hunk 6]
```diff
@@ -29,7 +30,15 @@ const styles = (theme) => ({
     marginRight: theme.spacing(3),
   },
   hide: {
-    display: 'none',
+    visibility: 'hidden',
+  },
+  toolbar: {
+    justifyContent: 'space-between',
+    padding: '0px',
```
Comment (pyphilia) [inline src/components/common/Header.js:37]: You can use `theme.spacing(0)` here.

File: src/langs/en.json [hunk 7]
```diff
@@ -7,6 +7,36 @@
...
+    "Load a File": "Load a File",
+    "Browse": "Browse",
+    "Submit": "Submit",
+    "Danger Zone! Proceed with caution as changes to this section might lead to data loss.": "Danger Zone! Proceed with caution as changes to this section might lead to data loss.",
+    "The database is empty.": "The database is empty.",
+    "Manually Edit the Database": "Manually Edit the Database",
+    "Use Sample Database": "Use Sample Database",
+    "Language": "Language",
+    "KB": "KB",
+    "Quick actions": "Actions rapides",
```
Comment (pyphilia) [inline src/langs/en.json:25]: French translation here.

File: src/reducers/settingsReducer.js [hunk 8]
```diff
@@ -0,0 +1,17 @@
+import { Map } from 'immutable';
+import { GET_LANGUAGE_SUCCESS, SET_LANGUAGE_SUCCESS } from '../types';
+import { DEFAULT_LANGUAGE } from '../config/constants';
+
+const INITIAL_STATE = Map({
+  lang: DEFAULT_LANGUAGE,
+});
+
+export default (state = INITIAL_STATE, { type, payload }) => {
+  switch (type) {
```
Comment (pyphilia) [inline src/reducers/settingsReducer.js:10]: Maybe add the flags dispatch here.

**Review by pyphilia — CHANGES_REQUESTED** [review]
Nice work! Some minor comments and it can be merged.

**Review by pyphilia — APPROVED** [review]

**Pull request merged**